File: 1999/6-6a.py
#!/usr/bin/python
if __name__ == "__main__":
	fin = file("prob6.in", "r")
	cnt = 1
	while 1:
		line = fin.readline()
		if line == "":
			break
		ln = line.split()
		caproom = int(ln[0])
		n = int(ln[1])
		conts = []
		for i in range(0, n):
			conts.append(int(ln[i + 2]))
		conts.sort()
		conts.reverse()
		ship = []
		while caproom >= conts[-1]:
			ship.append(conts[-1])
			caproom -= conts.pop()
		if caproom > 0:
			for i in conts:
				for j in ship:
					if caproom + j >= i and j <= i:
						caproom += j
						caproom -= i
						j = i
		print
		print "Ship", cnt
		print "------"
		print "Number of containers loaded:", len(ship)
		print "Weight under capacity:", caproom
		cnt += 1

File: 1999/6-6a.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
using namespace std;

void main()
{
	ifstream in("prob6.in");
	int caproom, cnt = 1, n, tmp;
	vector<int> conts;
	vector<int> ship;
	while (in >> caproom >> n)
	{
		for (int i = 0; i < n; i++)
		{
			in >> tmp;
			conts.push_back(tmp);
		}
		sort(conts.begin(), conts.end());
		reverse(conts.begin(), conts.end());
		while (caproom >= conts.back())
		{
			ship.push_back(conts.back());
			caproom -= conts.back();
			conts.pop_back();
		}
		if (caproom > 0)
			for (int i = 0; i < conts.size(); i++)
				for (int j = 0; j < ship.size(); j++)
					if ((caproom + ship[j] >= conts[i]) && (ship[j] < conts[i]))
					{
						caproom += ship[j];
						caproom -= conts[i];
						ship[j] = conts[i];
					}
		cout << "Ship " << cnt++ << endl << "------" << endl
			<< "Number of containers loaded: " << ship.size() << endl
			<< "Weight under capacity: " << caproom << endl << endl;
		ship.clear();
		conts.clear();
	}
}
